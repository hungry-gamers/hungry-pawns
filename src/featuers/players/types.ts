import type { PawnSize, MoveName } from '@/featuers/game/types.ts'

export type Player = {
  pawns: Record<PawnSize, number>
  capturedPawnsCounter: number
  arePawnsLocked: boolean
  specialMoves: MoveName[]
  skippedTurnsCount: number
}

export type PlayerPayload = { id: string; pawns: Player['pawns'] }

export type SpecialMoveStatus = 'not-allowed' | 'used'

export type PlayersStore = {
  players: Record<string, Player>
  pawnsStatus: 'locked' | 'picking'
}
