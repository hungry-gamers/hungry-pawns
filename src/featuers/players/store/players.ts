import { defineStore } from 'pinia'
import { reactive } from 'vue'
import * as PlayersT from '@/featuers/players/types.ts'
import * as PlayersService from '@/featuers/players/services/players.service.ts'
import * as GameT from '@/featuers/game/types.ts'
import {
  LAST_PENALTY_ON_SKIPPED_TURN_ON,
  PARALYZE_PROTECTION_IN_TURNS,
} from '@/featuers/players/services/players.service.ts'

export const usePlayersStore = defineStore('players', () => {
  const state = reactive<PlayersT.PlayersStore>({
    players: {},
    pawnsStatus: 'picking',
  })

  const createPlayers = (players: PlayersT.PlayerPayload[]) => {
    players.forEach((player) => {
      state.players[player.id] = PlayersService.createPlayer(player)
    })
  }

  const lockPawns = (playerId: string, pawns: Record<GameT.PawnSize, number>) => {
    if (state.players[playerId].arePawnsLocked) return
    state.players[playerId].pawns = { ...pawns }
    state.players[playerId].arePawnsLocked = true
    const locked = Object.values(state.players).filter(({ arePawnsLocked }) => arePawnsLocked)

    if (locked.length === getPlayers().length) {
      state.pawnsStatus = 'locked'
    }

    return state.pawnsStatus === 'locked'
  }

  const getCapturedPawnsCounter = (playerId: string) => {
    return state.players[playerId].capturedPawnsCounter
  }

  const getPawnsStatus = () => {
    return state.pawnsStatus
  }

  const getPlayers = (): string[] => {
    return Object.keys(state.players)
  }

  const getOpponentId = (currentPlayerId: string) => {
    const players = getPlayers()

    return players.find((id) => id !== currentPlayerId)!
  }

  const isPawnAvailableForPlayer = (playerId: string, pawnSize: GameT.PawnSize) => {
    return state.players[playerId].pawns[pawnSize]
  }

  const updateSkippedTurnsCount = (playerId: string) => {
    state.players[playerId].skippedTurnsCount += 1

    if (state.players[playerId].skippedTurnsCount > LAST_PENALTY_ON_SKIPPED_TURN_ON) {
      return LAST_PENALTY_ON_SKIPPED_TURN_ON
    }

    return state.players[playerId].skippedTurnsCount
  }

  const manipulatePawnAmount = (
    playerId: string,
    pawnSize: GameT.PawnSize,
    changeAmountBy: number,
  ) => {
    state.players[playerId].pawns[pawnSize] += changeAmountBy
  }

  const paralyzePlayer = (playerId: string) => {
    if (state.players[playerId].paralyzed.isActive) return

    state.players[playerId].paralyzed.isActive = true
    state.players[playerId].paralyzed.protectionExpiresIn =
      PlayersService.PARALYZE_PROTECTION_IN_TURNS
  }

  const canPlayerBeParalyzed = (playerId: string) => {
    return state.players[playerId].paralyzed.protectionExpiresIn === 0
  }

  const isPlayerParalyzed = (playerId: string) => {
    return state.players[playerId].paralyzed.isActive
  }

  const removeParalyzeEffect = (playerId: string) => {
    if (!state.players[playerId].paralyzed.isActive) return

    state.players[playerId].paralyzed.isActive = false
  }

  const decreaseParalyzeProtection = (playerId: string) => {
    if (state.players[playerId].paralyzed.protectionExpiresIn === 0) return
    state.players[playerId].paralyzed.protectionExpiresIn -= 1
  }

  const updatePlayerCapturedPawns = (playerId: string) => {
    state.players[playerId].capturedPawnsCounter++
  }

  const addSpecialMove = (playerId: string, moveName: GameT.MoveName) => {
    state.players[playerId].specialMoves.push(moveName)
  }

  const useSpecialMove = (
    playerId: string,
    moveName: GameT.MoveName,
  ): PlayersT.SpecialMoveStatus => {
    const index = state.players[playerId].specialMoves.indexOf(moveName)
    if (index === -1) return 'not-allowed'

    state.players[playerId].specialMoves.splice(index, 1)

    return 'used'
  }

  return {
    state,
    createPlayers,
    getPlayers,
    lockPawns,
    getPawnsStatus,
    getCapturedPawnsCounter,
    isPawnAvailableForPlayer,
    manipulatePawnAmount,
    updatePlayerCapturedPawns,
    removeParalyzeEffect,
    useSpecialMove,
    updateSkippedTurnsCount,
    getOpponentId,
    addSpecialMove,
    canPlayerBeParalyzed,
    decreaseParalyzeProtection,
    paralyzePlayer,
    isPlayerParalyzed,
  }
})
