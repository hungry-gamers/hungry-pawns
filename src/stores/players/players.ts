import { defineStore } from 'pinia'
import { reactive } from 'vue'
import type { PawnSize } from '@/stores/game/types.ts'

type Players = {
  pawns: Record<string, Record<PawnSize, number>>
  capturedPawnsCounter: Record<string, number>
}

export const usePlayersStore = defineStore('players', () => {
  const state = reactive<Players>({})
})
