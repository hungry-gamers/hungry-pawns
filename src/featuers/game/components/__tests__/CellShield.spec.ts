import { describe, expect, it, vi } from 'vitest'
import { mount } from '@vue/test-utils'
import CellShield from '@/featuers/game/components/CellShield.vue'
import { createTestingPinia } from '@pinia/testing'
import { usePlayersStore } from '@/featuers/players/store/players.ts'

describe('CellShield', () => {
  const pinia = createTestingPinia({
    createSpy: vi.fn,
  })

  const playersStore = usePlayersStore()
  playersStore.getPlayers = vi.fn(() => ['1', '2'])

  it('should match snapshot for small', () => {
    const wrapper = mount(CellShield, {
      props: { ownerId: '1' },
      global: {
        plugins: [pinia],
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
