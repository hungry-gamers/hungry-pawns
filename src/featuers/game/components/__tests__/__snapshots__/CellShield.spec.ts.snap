// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CellShield > should match snapshot for small 1`] = `
"<svg data-v-d3cbbe16="" class="shield-icon player-1-text" width="120" height="140" viewBox="0 0 100 120" xmlns="http://www.w3.org/2000/svg">
  <path data-v-d3cbbe16="" class="shield-base" d="M50 5 L90 25 V55 C90 85 70 105 50 115 C30 105 10 85 10 55 V25 Z"></path>
  <path data-v-d3cbbe16="" class="shield-gloss" d="M50 5 L90 25 V55 C90 75 70 95 50 105 C30 95 10 75 10 55 V25 Z"></path>
  <path data-v-d3cbbe16="" class="shield-highlight" d="M15 30 Q50 0 85 30"></path>
</svg>"
`;
